File: LeetCode/Explore/TopInterview/Amazon/Tree/236_lowest_common_ancestor_of_a_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root == None or root == p or root == q:
          return root
        
        lcal = self.lowestCommonAncestor(root.left, p, q)
        lcar = self.lowestCommonAncestor(root.right, p, q)
        
        if lcal and lcar:
          return root
        elif lcal:
          return lcal
        else:
          return lcar

File: LeetCode/Explore/TopInterview/MicroSoft/TreeGraph/236_lowest_common_ancestor_of_a_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
      if (!root || root == p || root == q) {
        return root;
      }
      TreeNode *lcal = lowestCommonAncestor(root->left, p, q);
      TreeNode *lcar = lowestCommonAncestor(root->right, p, q);
      if (lcal && lcar) {
        return root;
      }
      else if (lcal) {
        return lcal;
      }
      else {
        return lcar;
      }
    }
};
